src/jni/basic_file_reader.cpp

```diff
@@ -0,0 +1,68 @@
+#include <jni.h>
+#include "jni/basic_file_reader.h"
+
+#include <iostream>
+
+BasicFileReader::BasicFileReader(JNIEnv& env) : env(env) {
+  this->javaClass = env.FindClass("FileInputStreamMediator");
+  if (this->javaClass == nullptr) {
+    std::cerr << "ERROR: Java class FileInputStreamMediator not found" << std::endl;
+  }
+  this->javaConstructor = this->env.GetMethodID(this->javaClass, "<init>", "(Ljava/lang/String;)V");
+  // this would be a great place to verify the methods and arguments we expect to pass in are meeting our criteria
+  if (this->javaConstructor == nullptr) {
+    std::cerr << "ERROR: Java class FileInputStreamMediator does not comply with expected interface (ctor)" << std::endl;
+  }
+  this->javaClose = this->env.GetMethodID(this->javaClass, "close", "()V");
+  if (this->javaClose == nullptr) {
+    std::cerr << "ERROR: Java class FileInputStreamMediator does not comply with expected interface (close)" << std::endl;
+  }
+  // https://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/types.html#wp276 you'll need this. trust me.
```

Comment:
https://docs.oracle.com/en/java/javase/16/docs/specs/jni/types.html#type-signatures actually for the JNI version we're targeting